utilities/scale-json/src/lib.rs

```diff
@@ -0,0 +1,391 @@
+	}
+
+	fn visit_char<'scale, 'resolver>(
+		self,
+		value: char,
+		_type_id: scale_decode::visitor::TypeIdFor<Self>,
+	) -> Result<Self::Value<'scale, 'resolver>, Self::Error> {
+		Ok(serde_json::Value::String(value.to_string()))
+	}
+
+	fn visit_str<'scale, 'resolver>(
+		self,
+		value: &mut scale_decode::visitor::types::Str<'scale>,
+		_type_id: scale_decode::visitor::TypeIdFor<Self>,
+	) -> Result<Self::Value<'scale, 'resolver>, Self::Error> {
+		Ok(serde_json::Value::String(value.as_str()?.to_string()))
+	}
+
+	fn visit_bitsequence<'scale, 'resolver>(
+		self,
+		value: &mut scale_decode::visitor::types::BitSequence<'scale>,
+		_type_id: scale_decode::visitor::TypeIdFor<Self>,
+	) -> Result<Self::Value<'scale, 'resolver>, Self::Error> {
+		let s = value
+			.decode()?
+			.map(|r| r.map(|b| if b { '1' } else { '0' }))
+			.collect::<Result<String, _>>()?;
+		Ok(serde_json::Value::String(format!("0b{}", s)))
+	}
+
+	fn visit_composite<'scale, 'resolver>(
+		self,
+		value: &mut scale_decode::visitor::types::Composite<'scale, 'resolver, Self::TypeResolver>,
+		_type_id: scale_decode::visitor::TypeIdFor<Self>,
+	) -> Result<Self::Value<'scale, 'resolver>, Self::Error> {
+		if value.has_unnamed_fields() {
+			if value.remaining() == 1 {
+				// Simple wrapper struct. No need for array.
+				let field = value.next().unwrap()?;
+				assert!(field.name().is_none(), "Checked for unnamed fields.");
+				Ok(field.decode_with_visitor(RawJsonVisitor::<R>::default())?)
+			} else {
+				Ok(serde_json::Value::Array(
+					value
+						.map(|res| {
+							res.and_then(|field| {
+								let field_decoded =
+									field.decode_with_visitor(RawJsonVisitor::<R>::default())?;
+								if let Some(name) = field.name() {
+									Ok(serde_json::Value::Object(
+										std::iter::once((name.to_string(), field_decoded))
+											.collect(),
+									))
+								} else {
+									Ok(field_decoded)
+								}
+							})
+						})
+						.collect::<Result<Vec<_>, _>>()?,
+				))
+			}
+		} else {
+			Ok(serde_json::Value::Object(
+				value
+					.map(|res| {
+						res.and_then(|field| {
+							Ok((
+								field.name().expect("Checked for unnamed fields.").to_string(),
+								field.decode_with_visitor(RawJsonVisitor::<R>::default())?,
+							))
+						})
+					})
+					.collect::<Result<serde_json::Map<_, _>, _>>()?,
+			))
+		}
+	}
+
+	fn visit_variant<'scale, 'resolver>(
+		self,
+		value: &mut scale_decode::visitor::types::Variant<'scale, 'resolver, Self::TypeResolver>,
+		_type_id: scale_decode::visitor::TypeIdFor<Self>,
+	) -> Result<Self::Value<'scale, 'resolver>, Self::Error> {
+		if value.fields().remaining() == 0 {
+			Ok(serde_json::Value::String(value.name().to_string()))
+		} else {
+			Ok(serde_json::Value::Object(
+				std::iter::once(value)
+					.map(|value| {
+						Ok::<_, Self::Error>((
+							value.name().to_string(),
+							RawJsonVisitor::<R>::default()
+								.visit_composite(value.fields(), R::TypeId::default())?,
```

Comment:
What is the type id for in scale-decode?

Reply:
You can use it in the Visitor implementation's TypeResolver implementation to resolve additional information about a type. (See: https://docs.rs/scale-type-resolver/0.1.1/scale_type_resolver/trait.TypeResolver.html)

For example, the `PortableRegistry` type [implements this](https://docs.rs/scale-type-resolver/0.2.0/src/scale_type_resolver/portable_registry.rs.html#59), where type_id is just an integer and maps to metadata about the structure of the type. This metadata is used to delegate to the appropriate methods like visit_composite etc.

At *this* level of the decoding 'stack', we don't have access to the resolver so we can't resolve any information. The assumption is that we've reached the 'lowest resolution' and have to decide based on the available structural/primitive information.

Tbh. I'm not really sure why the type-id is passed without any means of resolving it... it does seem to defeat the purpose.